crates/sui-types/src/dynamic_field.rs

```diff
@@ -368,39 +369,212 @@ where
+    K: Serialize + fmt::Debug,
 {
-    let key_type_tag = K::get_type_tag(key_type_params);
-    get_dynamic_field_object_from_store_impl(object_store, parent_id, key, &key_type_tag)
+    pub fn object_id(&self) -> Result<ObjectID, SuiError> {
+        derive_dynamic_field_id(self.0, &self.2, &bcs::to_bytes(&self.1).unwrap())
+            .map_err(|e| SuiError::DynamicFieldReadError(e.to_string()))
+    }
+
+    pub fn into_id(self) -> Result<DynamicFieldID<K>, SuiError> {
+        let id = self.object_id()?;
+        Ok(DynamicFieldID::<K>::new(self.0.into(), id))
+    }
+
+    pub fn into_field<V>(self, value: V) -> Result<DynamicField<K, V>, SuiError>
+    where
+        V: Serialize + DeserializeOwned + MoveTypeTagTrait,
+    {
+        let id = self.object_id()?;
+        let field = Field::<K, V> {
+            id: UID::new(id),
+            name: self.1,
+            value,
+        };
+        let type_tag = TypeTag::Struct(Box::new(DynamicFieldInfo::dynamic_field_type(
+            self.2,
+            V::get_type_tag(),
+        )));
+        Ok(DynamicField(field, type_tag))
+    }
 }
 
-/// Get a dynamic field from the store with a generic key type. The key type may
-/// only have phantom type parameters, because it must have a fixed serialization
-/// format.
-pub fn get_dynamic_field_from_store_generic<K, V>(
-    object_store: &dyn ObjectStore,
-    parent_id: ObjectID,
-    key: &K,
-    key_type_params: &[TypeTag],
-) -> Result<V, SuiError>
+pub struct DynamicField<K, V>(Field<K, V>, TypeTag);
+
+impl<K, V> DynamicField<K, V>
 where
-    K: MoveTypeTagTraitGeneric + Serialize + DeserializeOwned + fmt::Debug,
-    V: Serialize + DeserializeOwned,
+    K: Serialize,
+    V: Serialize,
 {
-    let key_type_tag = K::get_type_tag(key_type_params);
-    get_dynamic_field_from_store_impl(object_store, parent_id, key, &key_type_tag)
+    // Do not call except in tests to avoid possible conservation bugs.
+    pub fn into_move_object_unsafe_for_testing(
+        self,
+        version: SequenceNumber,
+    ) -> Result<MoveObject, SuiError> {
+        let field = self.0;
+        let type_tag = self.1;
+        let TypeTag::Struct(struct_tag) = type_tag else {
+            unreachable!()
+        };
+        // TODO(address-balances): more efficient type repr
+        let move_object_type = MoveObjectType::from(*struct_tag);
+
+        let field_bytes =
+            bcs::to_bytes(&field).map_err(|e| SuiError::ObjectSerializationError {
+                error: e.to_string(),
+            })?;
+        Ok(unsafe {
+            MoveObject::new_from_execution_with_limit(
+                move_object_type,
+                false, // A dynamic field is never transferable, public or otherwise.
+                version,
+                field_bytes,
+                512,
+            )
+        }?)
+    }
+
+    pub fn into_inner(self) -> Field<K, V> {
+        self.0
+    }
+}
+
+pub struct DynamicFieldID<K: Serialize>(
```

Comment:
I don't fully understand this one. Could you add more doc comments? What does it represent exactly? Also not understanding the `load_object` function.

Reply:
Added a bunch more comments - but the best way to understand the use of this is to look at the uses of it in accumulator_root.rs. 

For instance, I started down this path when I noticed that the scheduler wants to read by ID, while everything else wants to read by parent + key. Also, in some cases I wanted a bounded/consistent read, and in some cases we don't care (metadata). So to do this just with functions we would need one set of functions for `_by_id` and one for `_by_parent_and_key`, times one set for `_consistent` and `_latest`, and so on.